cuda/matrix/ell_kernels.cu

```diff
@@ -128,16 +166,97 @@ void spmv(std::shared_ptr<const CudaExecutor> exec,
+            subwarp_size <<= 1;
+        }
+        if (subwarp_size == 32) {
+            nwarps_per_row =
+                std::min(ell_ncols / cuda_config::warp_size, nwarps / nrows);
+            nwarps_per_row = std::max(nwarps_per_row, 1);
+        }
+        if (nwarps_per_row > 1) {
+            atomic = true;
+        }
+    }
     const dim3 block_size(default_block_size, 1, 1);
-    const dim3 grid_size(ceildiv(a->get_size()[0], block_size.x),
-                         b->get_size()[1], 1);
-
-    abstract_spmv<<<grid_size, block_size, 0, 0>>>(
-        a->get_size()[0], as_cuda_type(a->get_const_values()),
-        a->get_const_col_idxs(), a->get_stride(),
-        a->get_num_stored_elements_per_row(),
-        as_cuda_type(b->get_const_values()), b->get_stride(),
-        as_cuda_type(c->get_values()), c->get_stride());
+    const dim3 grid_size(
+        ceildiv(nrows * subwarp_size * nwarps_per_row, block_size.x),
+        b->get_size()[1], 1);
+    switch (subwarp_size) {
+    case 1:
+        abstract_spmv<1><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 2:
+        abstract_spmv<2><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 4:
+        abstract_spmv<4><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 8:
+        abstract_spmv<8><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 16:
+        abstract_spmv<16><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 32:
+        if (atomic) {
+            ASSERT_NO_CUDA_ERRORS(
+                cudaMemset(c->get_values(), 0,
```

Comment:
There is a `zero_array` kernel. In general, I would not encourage directly calling cuda functions.